G-Research.OpenTelemetry.Processor.Partial/PartialActivityProcessor.cs

```diff
@@ -125,72 +162,119 @@ private ILoggerFactory CreateLoggerFactory()
             {
                 options.IncludeScopes = true;
-                options.AddProcessor(logProcessor);
+                options.AddProcessor(_logProcessor);
                 options.SetResourceBuilder(ResourceBuilder.CreateEmpty()
                     .AddAttributes(ParentProvider.GetResource().Attributes));
             });
         });
     }
 
-    public override void OnEnd(Activity data)
+    private Dictionary<string, object> GetHeartbeatLogRecordAttributes() => new()
+    {
+        ["partial.event"] = "heartbeat",
+        ["partial.frequency"] = $"{_heartbeatIntervalMilliseconds}ms",
+        ["partial.body.type"] = "json/v1",
+    };
+
+    private static Dictionary<string, object> GetStopLogRecordAttributes() => new()
+    {
+        ["partial.event"] = "stop",
+        ["partial.body.type"] = "json/v1",
+    };
+
+    private void ProcessQueues()
     {
-        using (logger.Value.BeginScope(GetStopLogRecordAttributes()))
+        var triggers = new WaitHandle[] { _shutdownTrigger };
+
+        while (true)
         {
-            logger.Value.LogInformation(
-                SpecHelper.Json(new TracesData(data, TracesData.Signal.Stop)));
+            try
+            {
+                WaitHandle.WaitAny(triggers, _processIntervalMilliseconds);
+
+                ProcessDelayedHeartbeatActivities();
+                ProcessReadyHeartbeatActivities();
+            }
+            catch (ObjectDisposedException)
+            {
+                return;
+            }
         }
+    }
+
+    private void ProcessDelayedHeartbeatActivities()
+    {
+        while (_delayedHeartbeatActivities.TryPeek(out var span) &&
+               span.InitialHeartbeatTime <= DateTime.UtcNow)
+        {
+            _delayedHeartbeatActivities.TryDequeue(out span);
 
-        endedActivities.Enqueue(new KeyValuePair<ActivitySpanId, Activity>(data.SpanId, data));
+            if (_activeActivities.TryGetValue(span.SpanId, out var activity))
+            {
+                using (_logger.Value.BeginScope(GetHeartbeatLogRecordAttributes()))
+                {
+                    _logger.Value.LogInformation(
+                        SpecHelper.Json(new TracesData(activity, TracesData.Signal.Heartbeat)));
+                }
+
+                _readyHeartbeatActivities.Enqueue((span.SpanId,
+                    DateTime.UtcNow.AddMilliseconds(_heartbeatIntervalMilliseconds)));
+            }
+        }
     }
 
-    protected override bool OnShutdown(int timeoutMilliseconds)
+    private void ProcessReadyHeartbeatActivities()
     {
-        try
+        while (_readyHeartbeatActivities.TryPeek(out var span) &&
+               span.NextHeartbeatTime <= DateTime.UtcNow)
         {
-            shutdownTrigger.Set();
+            _readyHeartbeatActivities.TryDequeue(out span);
+
+            if (!_activeActivities.TryGetValue(span.SpanId, out var activity))
+            {
+                continue;
+            }
```

Comment:
this is a fine pattern, but please make it consistent between `ProcessDelayedHeartbeatActivities` and `ProcessReadyHeartbeatActivities` by choosing one way or the other